File: time/src/main/java/com/tr/hsyn/time/Duration.java
package com.tr.hsyn.time;


import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;


/**
 * Allows specifying a <b><i>duration<i></b>.<br>
 * Durations are shown in specific units ({@link Unit}).<br><br>
 * <pre>
 *    var duration = Duration.of(Unit.MINUTE, 1981); // 1981 minutes duration
 *    //var duration = Duration.ofMinute(1981);// same as above
 * </pre>
 * <p>
 * There are no restrictions when specifying the duration,
 * the aforementioned minute is saved as it is.
 * So it is not treated as time, because if it were treated as time,
 * the minute value could be at most {@code 59}.<br>
 * However, a <b>duration</b> can be converted to an exact time equivalent.<br><br>
 *
 * <pre>
 * var duration = Duration.ofMinute(1981);
 * System.out.println(duration.toTimeDuration());
 * // [Duration{type=day, value=1}, Duration{type=hour, value=9}, Duration{type=minute, value=1}]
 * </pre>
 *
 * @see Unit
 * @see DurationGroup
 */
public class Duration {
	
	/**
	 * Time unit
	 */
	private final Unit unit;
	/**
	 * Time amount
	 */
	private final long value;
	
	/**
	 * Creates a new Duration.
	 *
	 * @param unit  Type of the time unit
	 * @param value Amount of the time
	 */
	Duration(@NotNull Unit unit, long value) {
		
		this.unit  = unit;
		this.value = value;
	}
	
	/**
	 * @return {@code true} if duration value is zero, {@code false} otherwise
	 */
	public boolean isZero() {
		
		return value == 0L;
	}
	
	/**
	 * @return {@code true} if duration value is not zero, {@code false} otherwise
	 */
	public boolean isNotZero() {
		
		return value != 0L;
	}
	
	/**
	 * Adds {@code value} to this {@link Duration}.
	 * İf the given value is negative, made subtract.
	 *
	 * <pre>
	 *     var duration = Duration.of(Unit.MINUTE, 5);
	 *     var duration2 = Duration.ofMinute(-1);
	 *     var duration3 = duration.plus(duration2); // Duration{type=minute, value=4}
	 *     var duration4 = duration.plus(-1L); // Duration{type=minute, value=4}
	 * </pre>
	 *
	 * @param value Amount of the time
	 * @return new {@link Duration}
	 */
	public Duration plus(long value) {
		
		return new Duration(this.unit, this.value + value);
	}
	
	/**
	 * Adds {@code Duration} to this {@link Duration}.
	 * İf the given duration is negative, made subtract.
	 * İf the given {@linkplain Duration} has a different {@linkplain Unit},
	 * it is converted to this {@linkplain Duration} unit.
	 * So returned {@link Duration} unit will be same as this {@linkplain Duration} unit.
	 *
	 * <pre>
	 *     var duration = Duration.of(Unit.MINUTE, 5);
	 *     var duration2 = Duration.ofMinute(-1);
	 *     var duration3 = duration.plus(duration2); // Duration{type=minute, value=4}
	 *     var duration4 = duration.plus(-1L); // Duration{type=minute, value=4}
	 * </pre>
	 *
	 * @param other Other {@linkplain Duration} object to add
	 * @return new {@link Duration}
	 */
	public Duration plus(@NotNull Duration other) {
		
		if (isDifferent(other)) {
			
			var d = other.getValueAs(this.unit);
			return new Duration(this.unit, this.value + d.getValue());
		}
		
		return new Duration(this.unit, this.value + value);
	}
	
	@Override
	public int hashCode() {
		
		return Objects.hash(unit, value);
	}
	
	/**
	 * Checks whether this <code>Duration</code> is equal to other {@link Duration}.
	 *
	 * @param o Other {@linkplain Duration} object
	 * @return {@code true} if {@link Unit} and {@link #value} are equal
	 */
	@Override
	public boolean equals(Object o) {
		
		return o instanceof Duration && unit.equals(((Duration) o).getUnit()) && value == ((Duration) o).getValue();
	}
	
	@NotNull
	@Override
	public String toString() {
		
		return toString("Duration{type=%2$s, value=%1$d}");
	}
	
	/**
	 * Returns formatted {@linkplain String} representation of this {@linkplain Duration}.
	 *
	 * @param formatted Formatted {@linkplain String} representation.
	 *                  For example: <code>"%d %s"</code>.
	 *                  The first parameter is the value, the second parameter is the unit.
	 *                  But can be changed the order of the parameters like this <code>"%2$s %1$d"</code> (as known).
	 * @return Formatted {@linkplain String} representation
	 */
	@NotNull
	public String toString(@NotNull String formatted) {
		
		return String.format(formatted, value, unit);
	}
	
	/**
	 * @return the unit of this {@linkplain Duration}
	 */
	public Unit getUnit() {
		
		return unit;
	}
	
	/**
	 * @return duration value of this {@linkplain Duration}
	 */
	public long getValue() {
		
		return value;
	}
	
	/**
	 * @param other Other {@linkplain Duration} object
	 * @return {@code true} if this <code>Duration</code> is not equal to other {@link Duration}.
	 */
	public boolean isDifferent(@NotNull Duration other) {
		
		return !this.equals(other);
	}
	
	/**
	 * Checks whether this <code>Duration</code> unit is different
	 * from given <code>Duration</code> {@link Unit}.
	 *
	 * @param other Other {@linkplain Unit}
	 * @return {@code true} if this <code>Duration</code> unit is different
	 * 		from given {@link Unit} by unit
	 */
	public boolean isDifferentByUnit(@NotNull Unit other) {
		
		return !unit.equals(other);
	}
	
	/**
	 * Determines whether this <code>Duration</code> unit is greater than given {@link Unit}.
	 *
	 * @param other Other {@linkplain Unit}
	 * @return {@code true} if this <code>Duration</code> unit is greater than given {@link Unit}
	 * @see Unit#isGreaterByUnit(Unit)
	 */
	public boolean isGreaterByUnit(@NotNull Unit other) {
		
		return unit.isGreaterByUnit(other);
	}
	
	/**
	 * Determines whether this <code>Duration</code> unit is less than given {@link Unit}.
	 *
	 * @param other Other {@linkplain Unit}
	 * @return {@code true} if this <code>Duration</code> unit is less than given {@link Unit}
	 * @see Unit#isLessByUnit(Unit)
	 */
	public boolean isLessByUnit(@NotNull Unit other) {
		
		return unit.isLessByUnit(other);
	}
	
	/**
	 * Determines whether this <code>Duration</code> is equal to given {@link Unit}.
	 *
	 * @param unit Type of the unit
	 * @return {@code true} if this <code>Duration</code> is equal to given {@link Unit}
	 */
	public boolean is(@NotNull Unit unit) {
		
		return this.unit.equals(unit);
	}
	
	/**
	 * Returns exactly how long the duration is as date format.
	 *
	 * <pre>
	 * var duration = Duration.ofMinute(1981);
	 * System.out.println(duration.toTimeDuration());
	 * // [Duration{type=day, value=1}, Duration{type=hour, value=9}, Duration{type=minute, value=1}]
	 * </pre>
	 *
	 * @return {@link DurationGroup}
	 */
	public DurationGroup toTimeDuration() {
		
		return Time.toDuration(toMilliseconds());
	}
	
	/**
	 * @return as milliseconds of this {@linkplain Duration}
	 */
	public long toMilliseconds() {
		//@off
		switch (unit) {
			case MILLISECOND: return value;
			case SECOND:      return value * 1000;
			case MINUTE:      return value * 60000;
			case HOUR:        return value * 3600000;
			case DAY:         return value * 86400000;
			case MONTH:       return value * 259200000;
			case YEAR:        return value * 36500000;
			
		}//@on
		
		throw new IllegalArgumentException("This is impossible : " + unit);
	}
	
	/**
	 * Converts the duration to the given unit
	 *
	 * @param unit Type of the unit
	 * @return {@link Duration}
	 */
	public Duration getValueAs(@NotNull Unit unit) {
		
		long value = toMilliseconds();
		//@off
		switch (unit) {
			case MILLISECOND: return new Duration(unit, value);
			case SECOND:      return new Duration(unit, value / 1000);
			case MINUTE:      return new Duration(unit, value / 60000);
			case HOUR:        return new Duration(unit, value / 3600000);
			case DAY:         return new Duration(unit, value / 86400000);
			case MONTH:       return new Duration(unit, value / 259200000);
			case YEAR:        return new Duration(unit, value / 36500000);
		}//@on
		
		throw new IllegalArgumentException("This is impossible : " + unit);
		
	}
	
	public static void main(String[] args) {
		
		var duration  = Duration.of(Unit.MINUTE, 5);
		var duration2 = Duration.ofMinute(-1);
		var duration3 = duration.plus(duration2); // Duration{type=minute, value=4}
		System.out.println(duration3.toString());
	}
	
	/**
	 * Creates a new Duration with given minutes.
	 *
	 * @param minutes Number of minutes
	 */
	@NotNull
	public static Duration ofMinute(long minutes) {
		
		return of(Unit.MINUTE, minutes);
	}
	
	/**
	 * Creates a new Duration.
	 *
	 * @param unit  Unit of time
	 * @param value Amount of time
	 * @return Duration
	 */
	@NotNull
	public static Duration of(@NotNull Unit unit, long value) {
		
		return new Duration(unit, value);
	}
	
	@NotNull
	public static DurationGroup of(@NotNull List<Duration> durations) {
		
		return new DurationGroup(durations);
	}
	
	@NotNull
	public static DurationGroup of(@NotNull Duration... durations) {
		
		return new DurationGroup(List.of(durations));
	}
	
	/**
	 * Creates a new Duration with value of zero.
	 *
	 * @param unit Unit of time
	 */
	@NotNull
	public static Duration ofZero(@NotNull Unit unit) {return new Duration(unit, 0L);}
	
	/**
	 * Creates a new Duration with given years.
	 *
	 * @param years Number of years
	 */
	@NotNull
	public static Duration ofYear(long years) {return of(Unit.YEAR, years);}
	
	/**
	 * Creates a new Duration with given months.
	 *
	 * @param months Number of months
	 */
	@NotNull
	public static Duration ofMonth(long months) {
		
		return of(Unit.MONTH, months);
	}
	
	/**
	 * Creates a new Duration with given days.
	 *
	 * @param days Number of days
	 */
	@NotNull
	public static Duration ofDay(long days) {return of(Unit.DAY, days);}
	
	/**
	 * Creates a new Duration with given hours.
	 *
	 * @param hours Number of hours
	 */
	@NotNull
	public static Duration ofHour(long hours) {
		
		return of(Unit.HOUR, hours);
	}
	
	/**
	 * Creates a new Duration with given seconds.
	 *
	 * @param seconds Number of seconds
	 */
	@NotNull
	public static Duration ofSecond(long seconds) {
		
		return of(Unit.SECOND, seconds);
	}
	
	/**
	 * Creates a new Duration with given milliseconds.
	 *
	 * @param milliseconds Number of milliseconds
	 **/
	@NotNull
	public static Duration ofMillisecond(long milliseconds) {
		
		return of(Unit.MILLISECOND, milliseconds);
	}
	
}

File: timek/src/main/java/tr/xyz/timek/duration/Duration.kt
package tr.xyz.timek.duration

import java.util.*
import kotlin.reflect.KProperty

/**
 * Represents a duration.
 *
 * Duration is an amount of time with certain unit (year, month, day, hour, minute, second,
 * millisecond). And this duration **is not** a time duration. So, the time units have no limit. For
 * example, can be defined a duration of a month with value `2_000`.
 *
 * ```
 *
 *    val _duration = Duration months 2000
 *    println(_duration) // 2000 months
 * ```
 *
 * And a duration can be converted to any unit.
 *
 * ```
 *
 *    val _duration = Duration months 2000 to TimeUnit.YEAR
 *    println(_duration) // 166 years
 *    println(_duration to TimeUnit.DAY) // 60590 days
 * ```
 *
 * If a duration is converted to bigger unit, it will be truncated.
 *
 * @constructor Creates a new duration
 * @property value value of the duration.
 * @property unit unit of the duration
 * @see TimeUnit
 */
class Duration(val value: Long, val unit: TimeUnit) {
	
	// region Properties
	/**
	 * `true` if the duration is zero
	 */
	fun isZero(): Boolean = value == 0L
	
	/**
	 * `true` if the duration is not zero
	 */
	fun isNotZero(): Boolean = value != 0L
	
	/**
	 * Millisecond equivalent of the current duration
	 */
	val asMilliseconds: Long by ValueConvertor(TimeUnit.MILLISECOND)
	
	/**
	 * Seconds equivalent of the current duration
	 */
	val asSeconds: Long by ValueConvertor(TimeUnit.SECOND)
	
	/**
	 * Minutes equivalent of the current duration
	 */
	val asMinutes: Long by ValueConvertor(TimeUnit.MINUTE)
	
	/**
	 * Hours equivalent of the current duration
	 */
	val asHours: Long by ValueConvertor(TimeUnit.HOUR)
	
	/**
	 * Days equivalent of the current duration
	 */
	val asDays: Long by ValueConvertor(TimeUnit.DAY)
	
	/**
	 * Months equivalent of the current duration
	 */
	val asMonths: Long by ValueConvertor(TimeUnit.MONTH)
	
	/**
	 * Years equivalent of the current duration
	 */
	val asYears: Long by ValueConvertor(TimeUnit.YEAR)
	
	// endregion
	
	// region private class ValueConvertor(private val unit: TimeUnit) {
	/**
	 * A utility class for converting durations between different time units.
	 *
	 * @property unit The target time unit for conversion.
	 */
	private class ValueConvertor(private val unit: TimeUnit) {
		
		operator fun getValue(ref: Duration, property: KProperty<*>): Long = when (unit) {
			
			TimeUnit.MILLISECOND -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value
				TimeUnit.SECOND      -> ref.value * 1000
				TimeUnit.MINUTE      -> ref.value * 1000 * 60
				TimeUnit.HOUR        -> ref.value * 1000 * 60 * 60
				TimeUnit.DAY         -> ref.value * 1000 * 60 * 60 * 24
				TimeUnit.MONTH       -> ref.value * 1000 * 60 * 60 * 24 * 30
				TimeUnit.YEAR        -> ref.value * 1000 * 60 * 60 * 24 * 365
			}
			
			TimeUnit.SECOND      -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 1000
				TimeUnit.SECOND      -> ref.value
				TimeUnit.MINUTE      -> ref.value * 60
				TimeUnit.HOUR        -> ref.value * 60 * 60
				TimeUnit.DAY         -> ref.value * 60 * 60 * 24
				TimeUnit.MONTH       -> ref.value * 60 * 60 * 24 * 30
				TimeUnit.YEAR        -> ref.value * 60 * 60 * 24 * 365
			}
			
			TimeUnit.MINUTE      -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 60000
				TimeUnit.SECOND      -> ref.value / 60
				TimeUnit.MINUTE      -> ref.value
				TimeUnit.HOUR        -> ref.value * 60
				TimeUnit.DAY         -> ref.value * 60 * 24
				TimeUnit.MONTH       -> ref.value * 60 * 24 * 30
				TimeUnit.YEAR        -> ref.value * 60 * 24 * 365
			}
			
			TimeUnit.HOUR        -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 3600000
				TimeUnit.SECOND      -> ref.value / 3600
				TimeUnit.MINUTE      -> ref.value / 60
				TimeUnit.HOUR        -> ref.value
				TimeUnit.DAY         -> ref.value * 24
				TimeUnit.MONTH       -> ref.value * 24 * 30
				TimeUnit.YEAR        -> ref.value * 24 * 365
			}
			
			TimeUnit.DAY         -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 86400000
				TimeUnit.SECOND      -> ref.value / 86400
				TimeUnit.MINUTE      -> ref.value / 1440
				TimeUnit.HOUR        -> ref.value / 24
				TimeUnit.DAY         -> ref.value
				TimeUnit.MONTH       -> ref.value * 30
				TimeUnit.YEAR        -> ref.value * 365
			}
			
			TimeUnit.MONTH       -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 2592000000
				TimeUnit.SECOND      -> ref.value / 2592000
				TimeUnit.MINUTE      -> ref.value / 43200
				TimeUnit.HOUR        -> ref.value / 720
				TimeUnit.DAY         -> ref.value / 30
				TimeUnit.MONTH       -> ref.value
				TimeUnit.YEAR        -> ref.value * 12
			}
			
			TimeUnit.YEAR        -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 31536000000
				TimeUnit.SECOND      -> ref.value / 31536000
				TimeUnit.MINUTE      -> ref.value / 525600
				TimeUnit.HOUR        -> ref.value / 8760
				TimeUnit.DAY         -> ref.value / 365
				TimeUnit.MONTH       -> ref.value / 12
				TimeUnit.YEAR        -> ref.value
			}
		}
	}
	// endregion
	
	// region Member functions
	/**
	 * Converts the duration to the given unit.
	 *
	 * @param unit [TimeUnit] convert to
	 * @return new [Duration] with the specified unit. If unit is bigger than the current unit, it
	 *     will be truncated
	 */
	infix fun to(unit: TimeUnit): Duration {
		return when (unit) {
			TimeUnit.MILLISECOND -> asMilliseconds milliseconds dDuration
			TimeUnit.SECOND      -> asSeconds seconds dDuration
			TimeUnit.MINUTE      -> asMinutes minutes dDuration
			TimeUnit.HOUR        -> asHours hours dDuration
			TimeUnit.DAY         -> asDays days dDuration
			TimeUnit.MONTH       -> asMonths months dDuration
			TimeUnit.YEAR        -> asYears years dDuration
		}
	}
	/**
	 * Returns a new [Duration] with the value and the same current unit.
	 *
	 * @param value The value of the new duration
	 * @return A new [Duration] object with the value and the same current unit
	 */
	fun newDuration(value: Long) = Duration(value, unit)
	// endregion
	
	// region Operator functions
	/**
	 * Adds the given [duration] to the current duration and returns a new [Duration] object.
	 *
	 * @param duration The duration to be added.
	 * @return A new [Duration] object representing the sum of the current duration and the given
	 *     duration.
	 */
	operator fun plus(duration: Long): Duration = newDuration(value + duration)
	/**
	 * Adds the given [duration] to the current duration and returns a new [Duration] object.
	 *
	 * @param duration The duration to be added.
	 * @return A new [Duration] object representing the sum of the current duration and the given
	 *     duration. If the given duration unit is different than the current duration unit, it will
	 *     be converted to the current duration unit and if it is bigger than the current unit, it
	 *     will be truncated.
	 */
	operator fun plus(duration: Duration): Duration = newDuration(value + (duration to unit).value)
	/**
	 * Subtracts the given [duration] to the current duration and returns a new [Duration] object.
	 *
	 * @param duration The duration to be subtracted.
	 * @return A new [Duration] object representing the minus of the current duration and the given
	 *     duration. If the given duration unit is different than the current duration unit, it will
	 *     be converted to the current duration unit and if it is bigger than the current unit, it
	 *     will be truncated.
	 */
	operator fun minus(duration: Duration): Duration = newDuration(value - (duration to unit).value)
	/**
	 * Subtracts the given [duration] from the current duration and returns a new [Duration] with the
	 * result.
	 *
	 * @param duration The duration to be subtracted
	 * @return A new [Duration] object representing the minus of the current duration and the given
	 *     duration.
	 */
	operator fun minus(duration: Long): Duration = newDuration(value - duration)
	// endregion
	
	// region Overrides functions 
	override fun toString() = "$value $unit"
	override fun equals(other: Any?): Boolean = other is Duration && value == other.value && unit == other.unit
	override fun hashCode() = Objects.hash(value, unit)
	// endregion
	
	// region Companion functions
	companion object {
		/**
		 * Returns a new [Duration] with the value and unit.
		 *
		 * @param value value
		 * @param unit [TimeUnit]
		 * @return new [Duration]
		 */
		fun of(value: Long, unit: TimeUnit) = Duration(value, unit)
		/**
		 * Returns a new [Duration] with the value in milliseconds.
		 *
		 * @param value value
		 * @return new [Duration]
		 */
		infix fun milliseconds(value: Long): Duration = Duration(value, TimeUnit.MILLISECOND)
		/**
		 * Returns a new [Duration] with the value in seconds.
		 *
		 * @param value value
		 * @return new [Duration]
		 */
		infix fun seconds(value: Long): Duration = Duration(value, TimeUnit.SECOND)
		/**
		 * Returns a new [Duration] with the value in minutes.
		 *
		 * @param value value
		 * @return new [Duration]
		 */
		infix fun minutes(value: Long): Duration = Duration(value, TimeUnit.MINUTE)
		/**
		 * Returns a new [Duration] with the value in hours.
		 *
		 * @param value value
		 * @return new [Duration]
		 */
		infix fun hours(value: Long): Duration = Duration(value, TimeUnit.HOUR)
		/**
		 * Returns a new [Duration] with the value in days.
		 *
		 * @param value value
		 * @return new [Duration]
		 */
		infix fun days(value: Long): Duration = Duration(value, TimeUnit.DAY)
		/**
		 * Returns a new [Duration] with the value in months.
		 *
		 * @param value value
		 * @return new [Duration]
		 */
		infix fun months(value: Long): Duration = Duration(value, TimeUnit.MONTH)
		/**
		 * Returns a new [Duration] with the value in years.
		 *
		 * @param value value
		 * @return new [Duration]
		 */
		infix fun years(value: Long): Duration = Duration(value, TimeUnit.YEAR)
	}
	// endregion
}




































